### PR TITLE
public/settings.spec.ts

### PR DESCRIPTION
**Pull request by shrshindeMSFT**
- Cleaning up the code 
- Adding new test cases for settings.spec.ts 
- Refactoring unit test cases

### CODE REVIEW

File: packages/teams-js/test/public/settings.spec.ts [hunk 1]
```diff
@@ -71,46 +57,131 @@ describe('settings', () => {
...
+    ];
+    const expectedSettings: settings.Settings = {
+      suggestedDisplayName: 'someSuggestedDisplayName',
+      contentUrl: 'someContentUrl',
+      websiteUrl: 'someWebsiteUrl',
+      entityId: 'someEntityId',
+    };
+
+    it('settings.getSettings should not allow calls before initialization', () => {
       const expectedSettings: settings.Settings = {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:75]: This is also defined on line 67-72. Should it be removed from within the test case?

File: packages/teams-js/test/public/settings.spec.ts [hunk 2]
```diff
@@ -71,46 +57,131 @@ describe('settings', () => {
...
+    });
 
-      utils.respondToMessage(message, expectedSettings);
+    Object.values(FrameContexts).forEach(context => {
+      if (!allowedContexts.some(allowedContexts => allowedContexts == context)) {
+        it(`settings.getSettings does not allow calls from ${context} context`, async () => {
+          await utils.initializeWithContext(context);
+          expect(() => {
+            settings.getSettings(settings => {
+              expect(settings).toBe(expectedSettings);
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:94]: I don't think this is correct. We don't want to validate that the settings are correct in the tests where we expect it to throw an error. I'm concerned that the tests are passing with both "expect success" and "expect error" in the same block of code. If we get to line 94, the test should fail.

File: packages/teams-js/test/public/settings.spec.ts [hunk 3]
```diff
@@ -71,46 +57,131 @@ describe('settings', () => {
...
+            `This call is only allowed in following contexts: ${JSON.stringify(
+              allowedContexts,
+            )}. Current context: "${context}".`,
+          );
+        });
+      } else {
+        it(`settings.getSettings should successfully get settings from ${context} context`, async () => {
+          await utils.initializeWithContext(context);
+          settings.getSettings(settings => {
+            expect(settings).toBe(expectedSettings);
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:106]: This check makes sense here in the success case.

File: packages/teams-js/test/public/settings.spec.ts [hunk 4]
```diff
@@ -71,46 +57,131 @@ describe('settings', () => {
...
+        });
+      }
     });
   });
 
-  it('should successfully register a save handler', () => {
-    utils.initializeWithContext('settings').then(() => {
+  describe('Testing settings.registerOnSaveHandler function', () => {
+    const allowedContexts = [FrameContexts.settings];
+    it('settings.registerOnSaveHandler should not allow calls before initialization', () => {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:160]: nit: newline before the start of the test case

File: packages/teams-js/test/public/settings.spec.ts [hunk 5]
```diff
@@ -71,46 +57,131 @@ describe('settings', () => {
...
+      if (!allowedContexts.some(allowedContexts => allowedContexts == context)) {
+        it(`settings.registerOnSaveHandler does not allow calls from ${context} context`, async () => {
+          await utils.initializeWithContext(context);
+          let handlerCalled = false;
+          expect(() => settings.registerOnSaveHandler(() => (handlerCalled = true))).toThrowError(
+            `This call is only allowed in following contexts: ${JSON.stringify(
+              allowedContexts,
+            )}. Current context: "${context}".`,
+          );
+        });
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:179]: You could add a check that handlerCalled is still false.

File: packages/teams-js/test/public/settings.spec.ts [hunk 6]
```diff
@@ -120,10 +191,10 @@ describe('settings', () => {
 
       expect(handlerCalled).toBe(true);
     });
-  });
 
-  it('should successfully register a remove handler', () => {
-    return utils.initializeWithContext('settings').then(() => {
+    it('settings.registerOnSaveHandler should successfully register a remove handler', async () => {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:195]: I know you didn't originally write this test, but what exactly is a "remove handler"?

File: packages/teams-js/test/public/settings.spec.ts [hunk 7]
```diff
@@ -138,120 +209,154 @@ describe('settings', () => {
...
-    expect(handlerCalled).toBe(true);
-    const message = utils.findMessageByFunc('settings.save.failure');
-    expect(message).not.toBeNull();
-    expect(message.args.length).toBe(1);
-    expect(message.args[0]).toBe('someReason');
+      let handlerCalled = false;
+      settings.registerOnSaveHandler(saveEvent => {
+        saveEvent.notifySuccess();
+        expect(() => saveEvent.notifySuccess()).toThrowError('The SaveEvent may only notify success or failure once.');
+        expect(() => saveEvent.notifyFailure()).toThrowError('The SaveEvent may only notify success or failure once.');
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:273]: I think another valid test case is doing the same sequence of actions except the first thing that is called is saveEvent.notifyFailure (like on line 271).

File: packages/teams-js/test/public/settings.spec.ts [hunk 8]
```diff
@@ -138,120 +209,154 @@ describe('settings', () => {
...
+          expect(() => settings.registerOnRemoveHandler(() => true)).toThrowError(
+            `This call is only allowed in following contexts: ${JSON.stringify(
+              allowedContexts,
+            )}. Current context: "${context}".`,
+          );
+        });
+      }
+    });
+
+    it('should successfully register a remove handler', async () => {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:301]: I don't think this test ('should successfully register a remove handler') is actually verifying anything?

File: packages/teams-js/test/public/settings.spec.ts (hunk 7 above)
Comment (erinha) [reply packages/teams-js/test/public/settings.spec.ts:273]: Can you add such a test case please?

File: packages/teams-js/test/public/settings.spec.ts (hunk 5 above)
Comment (erinha) [reply packages/teams-js/test/public/settings.spec.ts:179]: I think adding that check is worthwhile

File: packages/teams-js/test/public/settings.spec.ts (hunk 2 above)
Comment (erinha) [reply packages/teams-js/test/public/settings.spec.ts:94]: In fact, if you could add an expect that fails if this line of code is reached, that would be preferable.

File: packages/teams-js/test/public/settings.spec.ts [hunk 9]
```diff
@@ -138,120 +209,154 @@ describe('settings', () => {
...
-      handlerCalled = true;
+      expect(handlerCalled).toBe(true);
+      const message = utils.findMessageByFunc('settings.remove.failure');
+      expect(message).not.toBeNull();
+      expect(message.args.length).toBe(1);
+      expect(message.args[0]).toBe('someReason');
     });
+  });
 
+  it('should successfully notify success on save when there is no registered handler', async () => {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:356]: I don't think this test case belongs with the removeHandler test cases. I think it's better grouped at the very beginning in the describe("settings") test group.

File: packages/teams-js/test/public/settings.spec.ts [hunk 10]
```diff
@@ -20,48 +21,33 @@ describe('settings', () => {
...
-
-  it('should not allow calls from the wrong context', () => {
-    utils.initializeWithContext('content').then(() => {
-      expect(() => settings.setValidityState(true)).toThrowError(
-        'This call is only allowed in following contexts: ["settings","remove"]. Current context: "content".',
-      );
+  describe('Testing settings.setValidityState function', () => {
+    const allowedContexts = [FrameContexts.settings, FrameContexts.remove];
+    Object.values(FrameContexts).forEach(context => {
+      if (!allowedContexts.some(allowedContexts => allowedContexts == context)) {
```
Comment (erinha) [inline packages/teams-js/test/public/settings.spec.ts:27]: Please use "==="

File: packages/teams-js/test/public/settings.spec.ts (hunk 1 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts]: Removed duplicates.

File: packages/teams-js/test/public/settings.spec.ts (hunk 7 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts:292]: Added in the latest commit.

File: packages/teams-js/test/public/settings.spec.ts (hunk 9 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts]: Grouped to the top for the settings group.

File: packages/teams-js/test/public/settings.spec.ts (hunk 10 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts]: Modified in the latest commit.

File: packages/teams-js/test/public/settings.spec.ts (hunk 4 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts:165]: Added in the latest commit.

File: packages/teams-js/test/public/settings.spec.ts (hunk 2 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts]: I fix it by sending empty callback function can you please check this in the latest commit.

File: packages/teams-js/test/public/settings.spec.ts (hunk 6 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts]: It was issue with naming convention I have updated it.

File: packages/teams-js/test/public/settings.spec.ts [hunk 11]
```diff
@@ -6,6 +7,9 @@ describe('settings', () => {
   // Use to send a mock message from the app.
 
   const utils = new Utils();
+  const emptyCallback = (): void => {
```
Comment (jadahiya-MSFT) [inline packages/teams-js/test/public/settings.spec.ts:10]: Why is this needed?

File: packages/teams-js/test/public/settings.spec.ts [hunk 12]
```diff
@@ -71,46 +70,136 @@ describe('settings', () => {
...
+      }).toThrowError('The library has not yet been initialized');
+    });
 
-      const message = utils.findMessageByFunc('settings.getSettings');
-      expect(message).not.toBeNull();
+    Object.values(FrameContexts).forEach(context => {
+      if (!allowedContexts.some(allowedContexts => allowedContexts === context)) {
+        it(`settings.getSettings does not allow calls from ${context} context`, async () => {
+          await utils.initializeWithContext(context);
+          expect(() => settings.getSettings(emptyCallback)).toThrowError(
```
Comment (jadahiya-MSFT) [inline packages/teams-js/test/public/settings.spec.ts:99]: Can you verify if this works as intended? Sometimes `toThrowError` provides a positive test pass when the output isn't correct.

Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts:99]: Yes, I have verified the test case in both positive test pass and fail.

File: packages/teams-js/test/public/settings.spec.ts (hunk 11 above)
Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts:10]: We need this to send it as a parameter to `getSettings` function.

File: packages/teams-js/test/public/settings.spec.ts (hunk 12 above)
Comment (jadahiya-MSFT) [reply packages/teams-js/test/public/settings.spec.ts:99]: Awesome

File: packages/teams-js/test/public/settings.spec.ts (hunk 11 above)
Comment (jadahiya-MSFT) [reply packages/teams-js/test/public/settings.spec.ts:10]: does this not work?
```
getSettings( () => {}).toThrowError(...
```

Comment (shrshindeMSFT) [reply packages/teams-js/test/public/settings.spec.ts:10]: TypeScript doesn't allow and throws error.

**Review by jadahiya-MSFT — APPROVED** [review]

**Review by erinha — APPROVED** [review]
:shipit:
